src/Symfony/Component/Serializer/Normalizer/ArrayDenormalizer.php

```diff
@@ -72,8 +82,15 @@ public function supportsDenormalization($data, string $type, string $format = nu
             throw new BadMethodCallException(sprintf('The nested denormalizer needs to be set to allow "%s()" to be used.', __METHOD__));
         }
 
-        return str_ends_with($type, '[]')
-            && $this->denormalizer->supportsDenormalization($data, substr($type, 0, -2), $format, $context);
+        if (!str_ends_with($type, '[]')) {
+            return false;
+        }
+
+        if ($this->checkFirstElement) {
+            $data = 0 === \count($data) ? null : reset($data);
```

Comment:
This basically means that we would call the nested denormalizers with `$data === null` which is as odd for two reasons:

* Denormalizers are not used to getting called with `null` as data, so they might throw or return `false`.
* If the original `$data` is an empty array, `ArrayDenormalizer` would not call `denormalize()` on the nested denormalizers at all.

Since we're attempting to remove odd behevior, I'd like to get it right this time. 😉 

Suggestion: If the new mode is active _and_ `$data` is an empty array, we should return `true`.

Reply:
Totally agree of getting it right directly :). You're right, passing null might indeed be odd for denomalizers and may introduce other unwanted issues. And incase `$data` is empty this denormalizer will basically do nothing.

The only side effect would be, when an unsupported type is given and `[]` is passed by it will not throw an error. But when data is given (e.g. ["x", "y", "z"]) it will fail because it couldn't find a denormalizer for the type. 

Think this side effect is only preventable by deprecating/ removing `$data` from the method signature, so we don't have the issue of "what to pass when the array is empty?". But I don't have a clear picture if `$data` in `supportsDenormalization` has a use case that can't be solved in the `denormalize`.

And incase we might want to introduce BC for 7.0 we can consider this option like already discussed in the issue itself.

Reply:
I totally agree with @derrabus there. 
Actually, if the array is empty, there is no need to check the first element as it doesn't exist.
Therefore, why not just resume the "default" behavior when the array is empty?